llm/llm.go

```diff
@@ -40,57 +37,110 @@ func New(workDir, model string, adapters, projectors []string, opts api.Options)
 		return nil, err
 	}
 
-	if runtime.GOOS == "darwin" {
-		var requiredMemory int64
-		var f16Multiplier int64 = 2
-
-		switch ggml.ModelType() {
-		case "3B", "7B":
-			requiredMemory = 8 * format.GigaByte
-		case "13B":
-			requiredMemory = 16 * format.GigaByte
-		case "30B", "34B", "40B":
-			requiredMemory = 32 * format.GigaByte
-		case "47B":
-			requiredMemory = 48 * format.GigaByte
-		case "65B", "70B":
-			requiredMemory = 64 * format.GigaByte
-		case "180B":
-			requiredMemory = 128 * format.GigaByte
-			f16Multiplier = 4
-		}
+	if opts.NumCtx < 4 {
+		opts.NumCtx = 4
+	}
 
-		systemMemory := int64(memory.TotalMemory())
+	fmt.Println("size", ggml.Size)
+	fmt.Println("filetype", ggml.FileType())
+	fmt.Println("architecture", ggml.ModelFamily())
+	fmt.Println("type", ggml.ModelType())
+	fmt.Println("name", ggml.Name())
+	fmt.Println("embd", ggml.NumEmbed())
+	fmt.Println("head", ggml.NumHead())
+	fmt.Println("head_kv", ggml.NumHeadKv())
+	fmt.Println("gqa", ggml.NumGQA())
+
+	available, _ := gpu.CheckVRAM()
+
+	// For now assume filesize = model size
+	// TODO: use actual model size
+	requiredModel := ggml.Size
+
+	// fp16 k,v matrices each require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each
+	requiredKv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.NumLayers()) * int64(ggml.NumEmbed()) * int64(ggml.NumHeadKv()) / int64(ggml.NumHead())
+
+	// this amount is the overhead + tensors in memory
+	// TODO: get this from the llama.cpp's graph calcluations instead of
+	// guessing it's ~1/8th of the kv cache times gqa
+	// this is slightly (~25%) higher than what it is in reality
+	// TODO: this needs to account for batch size != 512
+	requiredAlloc := int64(ggml.NumGQA()) * requiredKv / 8
+
+	requiredTotal := requiredModel + requiredKv + requiredAlloc
+
+	fmt.Println("system memory ", available)
+	fmt.Println("required model", requiredModel)
+	fmt.Println("required kv", requiredKv)
+	fmt.Println("required alloc", requiredAlloc)
+	fmt.Println("required total", requiredTotal)
+
+	if opts.NumGPU == -1 {
+		opts.NumGPU = int(ggml.NumLayers()) + 1
+	}
 
-		if ggml.FileType() == "F16" && requiredMemory*f16Multiplier > systemMemory {
-			return nil, fmt.Errorf("F16 model requires at least %s of memory", format.HumanBytes(requiredMemory))
-		} else if requiredMemory > systemMemory {
-			return nil, fmt.Errorf("model requires at least %s of memory", format.HumanBytes(requiredMemory))
+	// decide how many layers to put on the GPU
+	if opts.NumGPU > 0 {
+		switch runtime.GOOS {
+		case "darwin":
+			if requiredTotal > available {
+				fmt.Println("not enough vram available, falling back to CPU only")
```

Comment:
```suggestion
				log.Printf("not enough vram available, %sMB required, falling back to CPU only", requiredTotal)
```
This is probably logged elsewhere, in that case ignore this comment